### PR TITLE
SAMA5 GMAC driver improvements: prevent txtimeout from always firing and fix txbuffer leak during high-volume sends

### PR DESCRIPTION
**Pull request by adamfeuer**
### Summary

* SAMA5 GMAC gigabit ethernet driver improvements:
  * prevent GMAC txtimeout from always firing; without the fix, the GMAC would reset every 60 seconds no matter what was happening.
  * fix txbuffer leak during high-volume sends
  * may not completely fix GMAC driver hang under extreme conditions, more testing is needed

### Impact

* None if you don't use the SAMA5 GMAC gigabit ethernet driver.
* SAMA5 GMAC users will notice the driver won't crash as quickly at high volume send rates with large packet sizes.
* I'm still not confident this completely solves all issues, but it definitely improves things.

### Limitations / TODO

* More testing under extreme conditions (high-volume TCP sends with large send sizes) is necessary.
* The SAMA5 EMAC 10/100 ethernet driver may also have the txtimeout problem, this PR does not fix it.

### Detail

* txtimeout fix was a correction to a typo in the GMAC Transmit Status Register address; the old code had the wrong address, so could never check the TSR correctly.
* txbuffer leak seems to be caused by the DMA driver sending multiple packets at once during high volume sends. This should not happen, but seems to imply some kind of race condition, since it occurs only under extreme conditions and not at regular intervals. The code partially fixes it. 
* Race conditions or resource depletion still causes NuttX to crash occasionally under extreme conditions.

### Testing

* Manual, using tcpblaster and a [custom Python tcpecho client](https://github.com/adamfeuer/echotools) that allows testing of many different TCP send scenarios.

### How To Verify

* Run tcpblaster
* Run tcpecho and use the above Python client to test various TCP send sizes and frequencies

### CODE REVIEW

File: arch/arm/src/sama5/sam_gmac.c [hunk 1]
```diff
@@ -1642,14 +1657,13 @@ static int sam_gmac_interrupt(int irq, void *context, FAR void *arg)
    *   one to this bit.
    */
 
-  tsr = sam_getreg(priv, SAM_GMAC_TSR_OFFSET);
+  tsr = sam_getreg(priv, SAM_GMAC_TSR);
```
Comment (adamfeuer) [inline arch/arm/src/sama5/sam_gmac.c]: This is the fix for the txtimeout always firing; the old code had the wrong register address here.

File: arch/arm/src/sama5/sam_gmac.c [hunk 2]
```diff
@@ -1337,8 +1347,9 @@ static void sam_txdone(struct sam_gmac_s *priv)
       up_invalidate_dcache((uintptr_t)txdesc,
                            (uintptr_t)txdesc + sizeof(struct gmac_txdesc_s));
 
-      /* Is this TX descriptor still in use? */
 
+      // Is this TX descriptor done transmitting? (SAMA5D36 datasheet, p. 934)
```
Comment (Ouss4) [inline arch/arm/src/sama5/sam_gmac.c]:
```suggestion
      /* Is this TX descriptor done transmitting? (SAMA5D36 datasheet, p. 934) */
      
```

File: arch/arm/src/sama5/sam_gmac.c [hunk 3]
```diff
@@ -1337,8 +1347,9 @@ static void sam_txdone(struct sam_gmac_s *priv)
...
+      // First TX descriptor in chain GMACTXD_STA_USED = 1
```
Comment (Ouss4) [inline arch/arm/src/sama5/sam_gmac.c]:
```suggestion
      /* First TX descriptor in chain GMACTXD_STA_USED = 1 */
      
```

File: arch/arm/src/sama5/sam_gmac.c [hunk 4]
```diff
@@ -690,6 +689,7 @@ static int sam_transmit(struct sam_gmac_s *priv)
 
   /* Check parameter */
 
+
```
Comment (Ouss4) [inline arch/arm/src/sama5/sam_gmac.c]: There already is a blank line above.
```suggestion
```

File: arch/arm/src/sama5/sam_gmac.c [hunk 5]
```diff
@@ -1,5 +1,4 @@
 /****************************************************************************
- * arch/arm/src/sama5/sam_gmac.c
```
Comment (Ouss4) [inline arch/arm/src/sama5/sam_gmac.c]: Why is this removed?

**Comment by patacongo** [comment]
There are too many coding standard problems.  You need to run the modified files through tools/nxstyle and correct the coding standard violations.

**Comment by patacongo** [comment]
There are too many coding standard problems.  You need to run the modified files through tools/nxstyle and correct the coding standard violations.

File: arch/arm/src/sama5/sam_gmac.c (hunk 2 above)
Comment (adamfeuer) [reply arch/arm/src/sama5/sam_gmac.c]: @Ouss4 will fix

File: arch/arm/src/sama5/sam_gmac.c (hunk 3 above)
Comment (adamfeuer) [reply arch/arm/src/sama5/sam_gmac.c]: @Ouss4 will fix

File: arch/arm/src/sama5/sam_gmac.c (hunk 4 above)
Comment (adamfeuer) [reply arch/arm/src/sama5/sam_gmac.c]: @Ouss4 will fix

File: arch/arm/src/sama5/sam_gmac.c (hunk 5 above)
Comment (adamfeuer) [reply arch/arm/src/sama5/sam_gmac.c]: @Ouss4 that's a mistake, will fix.

**Comment by adamfeuer** [comment]
@patacongo @Ouss4 I will run nxstyle and fix the problems and redo the code. Is there a guide to compiling and running nxstyle? Would you be willing to point me at it?

**Comment by xiaoxiang781216** [comment]
> @patacongo @Ouss4 I will run nxstyle and fix the problems and redo the code. Is there a guide to compiling and running nxstyle? Would you be willing to point me at it?

run tools/checkpatch.sh, which will compile the tools for you. You can pass -h to get the usage.

**Comment by adamfeuer** [comment]
@xiaoxiang781216 Thanks, I will do that!

**Comment by adamfeuer** [comment]
@Ouss4 @patacongo @xiaoxiang781216 Thanks for the improvements and help– I ran `tools/checkpatch.sh` against the two files in this PR, and made a bunch of cleanups... many of the `nxstyle`-suggested changes were code I didn't touch, but I cleaned them up anyway. 

In the `sam_gmac.h` file, there were a bunch of `nxstyle` errors from the previous code that I didn't fix... they were multiple instances of:

* error: Missing blank line after comment
* error: Mixed case identifier found

I didn't fix these because it seemed the suggestions would be out of character for the formatting in the file. But I'm willing to make the changes to get the file to the point that there are no `nxstyle` errors. What should I do?

Will you take a look at these changes and see what you think? Do you have further improvements?

File: arch/arm/src/sama5/sam_gmac.c [hunk 6]
```diff
@@ -1,5 +1,5 @@
...
+ * arch/arm/src/chip/sam_gmac.c
```
Comment (Ouss4) [inline arch/arm/src/sama5/sam_gmac.c]: arch/arm/src/chip/ is the simlink of arch/aarm/src/sama5
```suggestion
 * arch/arm/src/sama5/sam_gmac.c
```

**Comment by Ouss4** [comment]
@adamfeuer  Thanks for fixing all those complaints.
Header files that describe hardware registers tend to raise a lot of nxstyle issues.  Most of these issues are Missing blank line after comments or Long line warnings.
That format is used in maybe all header files of the sort, I don't know if we can fix all of them or just accept them as exceptions.

**Comment by adamfeuer** [comment]
@Ouss4 I fixed that file path header issue. As for the other nxstyle issues with the remaining warnings in the header that were there from before I touched this file– fixing them would add a lot of changes to this PR. One idea would be to create a second PR that fixes them. 

Let me know what you want me to do.

**Comment by Ouss4** [comment]
@adamfeuer Thanks for addressing everything.
I think this PR is ready. I'll merge it.

Fixing all the nxstyle complaints in some header files can be overwhelming.
If you think you can provide a PR for this file that would be welcomed.
